### DISCUSSION TITLE
Is it possible to error when a required field is missing after unmarshalling?

### DISCUSSION THREAD

**Discussion by rdbo** [pelletier/go-toml#930] [Questions] [answered]
I wanted to know if it's possible to error when a required field was not found during the unmarshall.
Example:
```go
type MyConfig struct {
	DatabaseUrl string `toml:"database_url,required"`
}

var MyConfig cfg
if err := toml.Unmarshall(configFileContent, &cfg), err != nil {
	// Handle missing required field error, or any other error
}
```
I don't know if this is possible, but it is easier to check a single error than checking if each individual field doesn't contain an invalid value. If it isn't possible, I say it would be a nice addition.

**Comment by pelletier** [comment, marked as answer]
Really sorry I missed this question!

There is no notion of required field in go-toml. In the spirit of keeping go-toml minimal, I think this feature is better implemented outside the library, for example: https://github.com/go-playground/validator. If you want to answer the question "is this value present in the toml document?" you have three options that come to mind:

- Unmarshal into a struct, and use a package like `validator` to check its result.
- Unmarshal into an `interface{}` and walk its value to see what was in the file.
- Use the unstable/parser API to iterate over each key of the document.

**Comment by rdbo** [comment]
For reference, I ended up doing my own validator by unmarshalling the fields as pointers, then checking if the pointers are `nil` for fields marked as `required`. Not sure if that's a good option, but worked for me
